File: src/android/app/src/main/java/com/singlevoid/caterina/data/photograph/Photograph.java
////////////////////////////////////////////////////////////////////////////////////////////////////
//                                      LICENSE                                                   //
////////////////////////////////////////////////////////////////////////////////////////////////////
//                                                                                                //
// Copyright [2020] [Joan Albert Espinosa Muns]                                                   //
//                                                                                                //
// Licensed under the Apache License, Version 2.0 (the "License")                                 //
// you may not use this file except in compliance with the License.                               //
// You may obtain a copy of the License at                                                        //
//                                                                                                //
// http://www.apache.org/licenses/LICENSE-2.0                                                     //
//                                                                                                //
// Unless required by applicable law or agreed to in writing, software                            //
// distributed under the License is distributed on an "AS IS" BASIS,                              //
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.                       //
// See the License for the specific language governing permissions and                            //
// limitations under the License.                                                                 //
//                                                                                                //
////////////////////////////////////////////////////////////////////////////////////////////////////

package com.singlevoid.caterina.data.photograph;

import android.content.Context;
import android.content.Intent;
import android.location.Location;
import android.net.Uri;
import android.os.Parcel;
import android.os.Parcelable;

import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.singlevoid.caterina.R;
import com.singlevoid.caterina.data.map.MapLocation;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Objects;

public class Photograph implements Parcelable{


    ////////////////////////////////////////////////////////////////////////////////////////////////
    //                                          VARIABLES                                         //
    ////////////////////////////////////////////////////////////////////////////////////////////////


    private String id;
    private String title;
    private String description;
    private String author;
    private String source;
    private MapLocation location;
    public PhotographDate date;
    private ArrayList<String> tags;
    private double distance;
    private String license;

    ////////////////////////////////////////////////////////////////////////////////////////////////
    //                                    CONSTRUCTORS AND OVERRIDES                              //
    ////////////////////////////////////////////////////////////////////////////////////////////////


    public Photograph() {}


    public Photograph(String title, String description, String author, String source,
                      MapLocation location, PhotographDate date, ArrayList<String> tags,
                      String license) {
        setTitle(title);
        setDescription(description);
        setAuthor(author);
        setSource(source);
        setLocation(location);
        setDate(date);
        setTags(tags);
        setLicense(license);
    }


    ////////////////////////////////////////////////////////////////////////////////////////////////
    //                             SETTERS AND GETTERS                                            //
    ////////////////////////////////////////////////////////////////////////////////////////////////


    public String getId() {
        return id;
    }


    public void setId(String id) {
        this.id = id;
    }


    public String getTitle() {
        return title;
    }


    public void setTitle(String title) {
        this.title = title;
    }


    public String getDescription() {
        return description;
    }


    public void setDescription(String description) {
        this.description = description;
    }


    public String getAuthor() {
        return author;
    }


    public void setAuthor(String author) {
        this.author = author;
    }


    public String getSource() {
        return source;
    }


    public void setSource(String source) {
        this.source = source;
    }


    public MapLocation getLocation() {
        return location;
    }


    public void setLocation(MapLocation location) {
        this.location = location;
    }


    public PhotographDate getDate() {
        return date;
    }


    public void setDate(PhotographDate date) {
        this.date = date;
    }


    public ArrayList<String> getTags() {
        return tags;
    }


    public void setTags(ArrayList<String> tags) {
        this.tags = tags;
    }


    public double getDistance() {
        return distance;
    }


    public String getLicense() {
        return license;
    }


    public void setLicense(String license) {
        this.license = license;
    }


    public String getHumanDate(){
        if(isDated())   { return getDate().toString(); }
        else            { return null; }
    }


    public float getLatitude() {
        return location.getLat();
    }


    public float getLongitude() {
        return location.getLng();
    }


    public StorageReference lowQualityReference(){
        return FirebaseStorage.getInstance().getReference("low/" + getId());
    }


    public StorageReference highQualityReference(){
        return FirebaseStorage.getInstance().getReference("high/" + getId());
    }


    public String getWebUrl(@NotNull Context context){
        String url = context.getResources().getString(R.string.base_url);
        return url + "/photograph/" + getId();
    }


    ////////////////////////////////////////////////////////////////////////////////////////////////
    //                                    PARCELABLE                                              //
    ////////////////////////////////////////////////////////////////////////////////////////////////


    protected Photograph(@NotNull Parcel in) {
        id = in.readString();
        title = in.readString();
        description = in.readString();
        author = in.readString();
        source = in.readString();
        location = in.readParcelable(MapLocation.class.getClassLoader());
        date = in.readParcelable(PhotographDate.class.getClassLoader());
        tags = in.createStringArrayList();
        distance = in.readDouble();
        license = in.readString();
    }


    public static final Creator<Photograph> CREATOR = new Creator<Photograph>() {
        @NotNull
        @Contract("_ -> new")
        @Override
        public Photograph createFromParcel(Parcel in) {
            return new Photograph(in);
        }

        @Override
        public Photograph[] newArray(int size) {
            return new Photograph[size];
        }
    };


    @Override
    public int describeContents() {
        return 0;
    }


    @Override
    public void writeToParcel(@NotNull Parcel parcel, int i) {
        parcel.writeString(id);
        parcel.writeString(title);
        parcel.writeString(description);
        parcel.writeString(author);
        parcel.writeString(source);
        parcel.writeParcelable(location, i);
        parcel.writeParcelable(date, i);
        parcel.writeStringList(tags);
        parcel.writeDouble(distance);
        parcel.writeString(license);
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////
    //                                          BOOLEANS                                          //
    ////////////////////////////////////////////////////////////////////////////////////////////////


    public boolean isLocalized() {
        return !Objects.isNull(location);
    }


    public boolean isDated() {
        return !Objects.isNull(date);
    }


    ////////////////////////////////////////////////////////////////////////////////////////////////
    //                                   METHODS                                                  //
    ////////////////////////////////////////////////////////////////////////////////////////////////


    public void setDistanceTo(@NotNull Location point) {
        final int R = 6371;
        double latDistance = Math.toRadians(point.getLatitude() - location.getLat());
        double lonDistance = Math.toRadians(point.getLongitude() - location.getLng());
        double a = Math.sin(latDistance / 2) * Math.sin(latDistance / 2)
                + Math.cos(Math.toRadians(location.getLat())) * Math.cos(Math.toRadians(point.getLatitude()))
                * Math.sin(lonDistance / 2) * Math.sin(lonDistance / 2);
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        distance = R * c * 1000;
    }


    public void SharePhotograph(Context context){
        Intent sendIntent = new Intent();
        sendIntent.setAction(Intent.ACTION_SEND);
        sendIntent.putExtra(Intent.EXTRA_TEXT, getWebUrl(context));
        sendIntent.setType("text/plain");
        context.startActivity(Intent.createChooser(sendIntent, null));
    }


    public void openInBrowser(@NotNull Context context){
        context.startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(getWebUrl(context))));
    }
}

File: src/webpage/src/app/model/photograph.ts

/** Copyright [2020] [Joan Albert Espinosa Muns]
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*   http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import { Injectable } from "@angular/core";
import { AngularFireStorage } from "@angular/fire/storage";
import { MapLocation } from "./map-location";
import { ObjectDate } from "./object-date";
import { RangedDate } from "./ranged-date";

@Injectable({providedIn: 'root'})
export class Photograph {

  private id: string = null;
  private title: string = null;
  private author: string = null;
  private description: string = null;
  private location: MapLocation;
  private date: ObjectDate;
  private tags: string[] = [];
  private license: string = null;
  private source: string = null;
  private lowQualityImage: string = null;
  private highQualityImage: string = null;

  private isDownloading: boolean = false;

  constructor(data?: any){
    this.setData(data)
  }

  public async setData(data: any){
    this.id = data?.id ? data.id : this.id;
    this.title = data?.title ? data.title : this.author;
    this.author = data?.author ? data.author : this.author;
    this.description = data?.description ? data.description : this.description;
    this.tags = data?.tags ? data.tags : this.tags;
    this.license = data?.license ? data.license : this.license;
    this.source = data?.source ? data.source : this.source;

    this.location = new MapLocation(data?.location)

    let year = new RangedDate(data?.date?.year)
    let month = new RangedDate(data?.date?.month)
    let day = new RangedDate(data?.date?.day)
    this.date = new ObjectDate(year, month, day)
  }


  public getId(): string{
    return this.id
  }


  public getSource(): string{
    return this.source
  }

  public getAuthor(): string{
    return this.author
  }


  public getTitle(): string{
    return this.title
  }


  public getDescription(): string{
    return this.description
  }


  public getTags(): string[]{
    return this.tags
  }


  public getDate(): ObjectDate{
    return this.date
  }


  public isGeolocalized(): boolean{
    return this.location.isGeolocalized()
  }


  public getLocation(): MapLocation{
    return this.location
  }


  public getLicense(): string{
    return this.license
  }


  public get isDated(): boolean{
      return (this.date.isDated)
  }


  public getLowQuality(storage?: AngularFireStorage){
    if(this.lowQualityImage){
      return this.lowQualityImage
    }
    else{
      this.downloadLowQualityImage(storage)
    }
  }


  public getHighQuality(storage: AngularFireStorage){
    if(this.highQualityImage){
      return this.highQualityImage
    }
    else{
      this.downloadHighQualityImage(storage)
    }
  }


  private async downloadLowQualityImage(storage: AngularFireStorage): Promise<void>{
    if(!this.isDownloading){
      this.isDownloading = true;
      storage.ref(`low/${this.id}`)
      .getDownloadURL()
      .toPromise().then(
        (url: string) => {
          this.isDownloading = false;
          this.lowQualityImage = url
        })
    }
  }


  private async downloadHighQualityImage(storage: AngularFireStorage): Promise<void>{
    if(!this.isDownloading){
      this.isDownloading = true;
      storage.ref(`high/${this.id}`)
      .getDownloadURL()
      .toPromise().then(
        (url: string) => {
          this.isDownloading = false;
          this.highQualityImage = url
        })
      }
    }


  public JSON(): object{

    var data = {}
    if ( this.getTitle() )              { data["title"] = this.getTitle() }
    if ( this.getDescription() )        { data["description"] = this.getDescription() }
    if ( this.getDate().toJSON() )      { data["date"] = this.getDate().toJSON() }
    if ( this.getLocation().toJSON() )  { data["location"] = this.getLocation().toJSON() }
    if ( this.getLicense() )            { data["license"] = this.getLicense() }
    if ( this.getTags() )               { data["tags"] = this.getTags() }
    if ( this.getSource() )             { data["source"] = this.getSource() }
    if ( this.getAuthor() )             { data["author"] = this.getAuthor() }

    return data;

  }


}
